File: app/src/main/java/com/example/administrator/classromapplication/view/ui/RegisteredActivity.java
package com.example.administrator.classromapplication.view.ui;

import android.view.View;
import android.widget.EditText;

import com.example.administrator.classromapplication.R;
import com.example.administrator.classromapplication.databinding.ActivityRegisteredBinding;
import com.example.administrator.classromapplication.viewmodel.UserViewModel;
import com.tool.util.DataUtils;
import com.tool.util.ToastHelp;
import com.tool.util.widget.CustomTitleBar;

import cn.bmob.v3.BmobUser;
import cn.bmob.v3.exception.BmobException;
import cn.bmob.v3.listener.LogInListener;
import cn.bmob.v3.listener.SaveListener;

public class RegisteredActivity extends BaseActivity<ActivityRegisteredBinding, UserViewModel> {
    @Override
    protected void initView() {
        ((CustomTitleBar) binding.titleBar).setTitle("注册");
    }

    @Override
    protected void initListener() {
        binding.btnRegist.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (checkETNotNull(binding.etUserName, "请输入用户名") && checkETNotNull(binding.etPwd, "请输入密码")
                        && checkETNotNull(binding.etPwdConfirm, "请输入确认密码") && checkPwdConfirm() && checkETNotNull(binding.etName, "请输入您的名字")) {
                    viewModel.setRoot(false);
                    viewModel.signUp(new SaveListener<UserViewModel>() {
                        @Override
                        public void done(UserViewModel o, BmobException e) {
                            if (e == null) {
                                ToastHelp.showToast("注册成功，开始登录");
                                BmobUser.loginByAccount(o.getUserName(), o.getPwd(), new LogInListener<Object>() {
                                    @Override
                                    public void done(Object o, BmobException e) {
                                        if (e == null) {
                                            goActivity(MainActivity.class, null);
                                        } else {
                                            ToastHelp.showToast("登录失败" + e.getMessage());
                                        }
                                    }
                                });
                            } else {
                                ToastHelp.showToast("注册失败" + e.getMessage());
                            }
                        }
                    });
                }
            }
        });
    }

    private boolean checkETNotNull(EditText editText, String errorMsg) {
        if (DataUtils.checkStrNotNull(editText.getText().toString())) return true;
        ToastHelp.showToast(errorMsg);
        return false;
    }

    private boolean checkPwdConfirm() {
        return binding.etPwd.getText().toString().trim().equals(binding.etPwdConfirm.getText().toString().trim());
    }

    @Override
    protected int getLayoutId() {
        return R.layout.activity_registered;
    }

    @Override
    protected UserViewModel getViewModel() {
        UserViewModel model = new UserViewModel();
        binding.setData(model);
        return model;
    }
}

File: app/src/main/java/com/example/administrator/classromapplication/view/ui/RegisteredActivity.kt
package com.example.administrator.classromapplication.view.ui

import android.widget.EditText
import cn.bmob.v3.exception.BmobException
import cn.bmob.v3.listener.SaveListener
import com.example.administrator.classromapplication.LoginUtil
import com.example.administrator.classromapplication.R
import com.example.administrator.classromapplication.databinding.ActivityRegisteredBinding
import com.example.administrator.classromapplication.viewmodel.UserViewModel
import com.tool.util.DataUtils
import com.tool.util.ToastHelp
import com.tool.util.widget.CustomTitleBar
import kotlinx.android.synthetic.main.activity_registered.*

/**
 * Created by Ervin on 2018/1/10.
 */
class RegisteredActivity : BaseActivity<ActivityRegisteredBinding, UserViewModel>() {
    override fun getViewModel(): UserViewModel {
        val model = UserViewModel()
        binding.data = model
        return model
    }

    override fun initView() {
        (titleBar as CustomTitleBar).setTitle("注册")
    }

    override fun initListener() {
        btn_regist.setOnClickListener {
            if (checkParam()) {
                showLoading()
                viewModel.setRoot(false)
                viewModel.signUp(object : SaveListener<UserViewModel>() {
                    override fun done(model: UserViewModel?, e: BmobException?) {
                        if (e == null) {
                            ToastHelp.showToast("注册成功，开始登录")
                            LoginUtil.login(model?.userName, model?.pwd, this@RegisteredActivity)
                        } else {
                            ToastHelp.showToast("注册失败" + e.message)
                        }
                    }
                })
            }
        }
    }

    fun checkEtnNotNull(editText: EditText, errorMsg: String): Boolean {
        if (DataUtils.checkStrNotNull(editText.text.toString())) return true
        ToastHelp.showToast(errorMsg)
        return false
    }

    fun checkParam(): Boolean {
        return checkEtnNotNull(binding.etUserName, "请输入用户名")
                && checkEtnNotNull(binding.etPwd, "请输入密码")
                && checkEtnNotNull(binding.etPwdConfirm, "请输入确认密码")
                && checkPwdConfirm()
                && checkEtnNotNull(binding.etName, "请输入您的名字")
    }

    fun checkPwdConfirm(): Boolean {
        return et_pwd.text.toString().trim() == et_pwd.text.toString().trim()
    }

    override fun getLayoutId(): Int {
        return R.layout.activity_registered
    }
}